src/core/highlight.js

```diff
@@ -45,32 +46,45 @@ export async function run(conf) {
         console.error("Timed-out waiting for highlight:", element);
         done();
       }, 4000);
-      const msg = {
-        action: "highlight",
-        code: element.textContent,
-        id: `highlight:${i}`,
-        languages: getLanguageHint(element.classList),
-      };
-      worker.addEventListener("message", function listener(ev) {
-        const {
-          data: { id, language, value },
-        } = ev;
-        if (id !== msg.id) {
-          return; // not for us!
-        }
-        element.innerHTML = value;
-        if (element.localName === "pre") {
-          element.classList.add("hljs");
-        }
-        if (language) {
-          element.classList.add(language);
-        }
-        clearTimeout(timeoutId);
-        worker.removeEventListener("message", listener);
-        done();
+      const elementsToHighlight =
+        element.querySelectorAll("code").length > 0
+          ? Array.from(element.querySelectorAll("code"))
+          : [element];
+      elementsToHighlight.map(elem => {
+        const msg = {
+          action: "highlight",
+          code: elem.innerText,
+          id: `highlight:${id}`,
+          languages: getLanguageHint(elem.classList),
+        };
+        worker.addEventListener("message", function listener(ev) {
+          const {
+            data: { id, language, value },
+          } = ev;
+          if (id !== msg.id) {
+            return; // not for us!
+          }
+          const lang = language !== undefined ? language : msg.languages[0];
+          switch (elem.localName) {
+            case "pre":
+              elem.innerHTML = `<code class="${lang}">${value}</code>`;
+              elem.classList.remove(lang);
+              elem.firstChild.classList.add("hljs");
+              break;
+            case "code":
+              elem.innerHTML = value;
+              elem.classList.add(language);
+              elem.classList.add("hljs");
+              break;
+          }
+          clearTimeout(timeoutId);
+          worker.removeEventListener("message", listener);
+          done();
+        });
+        element.setAttribute("aria-busy", "true");
+        worker.postMessage(msg);
       });
-      element.setAttribute("aria-busy", "true");
-      worker.postMessage(msg);
+      id++;
```

Comment:
this is going to be wrong :) imagine you have 2+ `<code>`s, they are going to get the same message id. You should increment the id on line 54 instead (before the `cont msg =`).

Reply:
Ah! my bad! I had to add on the line above i.e below `worker.postMessage`

Reply:
I think there might be a larger problem with the code tho... the `promisesToHighlight` were intended to only handle one highlight request at time... but `code` blocks now do multiple post messages... but resolve done() after one message 😱. 

I think I need to jump into this code a bit and refactor some of that because we are going to get weird async errors like "aria-busy" will be false before a second `code` update happens.

Reply:
could we not `element.setAttribute("aria-busy", "false");` set this for `elem` instead and at the end of each loop set it to `true` for the `elem` itself instead of `element` ?